### PR TITLE
Fixes #29092 - Set Active Tab for AuthSourceExternal form

### PR DESCRIPTION
**Pull request by aaditi-jain**
Bug:
https://projects.theforeman.org/issues/29092

Steps to reproduce bug:
1. log in to the dashboard as a administrator user   
1. On the dashboard, click on 'Administrator' in the left nav-bar and then click on 'Authentication sources'  
1. Then on the card titled as 'External', click on the icon at top-right  
1. Click on the 'edit' option from the dropdown
1. You will notice that on the page that loads (/auth_source_externals/{:id}/edit), there is no active tab and the tab contents are empty

Before Fix Screenshot:

29092_before

After Fix Screenshot:

29092 _after

### CODE REVIEW

**Comment by rahulbajaj0509** [comment]
[test foreman]

**Comment by rabajaj0509** [comment]
[test foreman]

**Comment by tbrisker** [comment]
OK to test

**Review by rabajaj0509 — DISMISSED** [review]
I have tested the PR and it works as expected! LGTM, will still wait for final review from @sharvit  @tbrisker . Meanwhile you might want to rebase your PR :)

File: app/views/auth_source_externals/_form.html.erb
Comment (ezr-ondrej) [inline app/views/auth_source_externals/_form.html.erb]:
```suggestion
      <li class='active'>
```

File: app/views/auth_source_externals/_form.html.erb
Comment (ezr-ondrej) [inline app/views/auth_source_externals/_form.html.erb]:
```suggestion
      <li class="<%= 'active' unless show_location_tab? %>">
```

File: app/views/auth_source_externals/_form.html.erb
Comment (ezr-ondrej) [inline app/views/auth_source_externals/_form.html.erb]:
```suggestion
      :active_tab => show_location_tab? ? :location : :organization %>
```

File: app/views/taxonomies/_loc_org_tabs.html.erb
Comment (ezr-ondrej) [inline app/views/taxonomies/_loc_org_tabs.html.erb:3]:
```suggestion
<% active_tab ||= nil %>
```

File: app/views/taxonomies/_loc_org_tabs.html.erb
Comment (ezr-ondrej) [inline app/views/taxonomies/_loc_org_tabs.html.erb]:
```suggestion
  <div class="tab-pane<%= ' active' if active_tab == :location %>" id="locations">
```

File: app/views/taxonomies/_loc_org_tabs.html.erb
Comment (ezr-ondrej) [inline app/views/taxonomies/_loc_org_tabs.html.erb]:
```suggestion
  <div class="tab-pane<%= ' active' if active_tab == :organization %>" id="organizations">
```

**Review by ezr-ondrej — COMMENTED** [review]
I don't think the helper is necessary, I've left idea how we could aproach it without the helper.
I'm not against the helper though :+1:
If you feel like you would like to have the helper, just please keep the formating
- <%=(space!) ... (space!)
- even though it's unfortunate, we use `=>` hash style
- (nitpick) html should use `"` (double quotes) e.g. `class="<%= .. %>"`

**Comment by aaditi-jain** [comment]
@rahulbajaj0509 @ezr-ondrej : Thank you for the review :-)
@ezr-ondrej : The approach that you suggested in the review seemed more readable to me. However (imho) using the helper will help keep the logic extracted at one place and view files neater. Hence continued to use the helper.
I updated the hash syntax and added the missing spaces as suggested.

**Review by kgaikwad — CHANGES_REQUESTED** [review]
It's good to have helper methods to extract the complex logic out of the view. With the current changes, I didn't find any good reason to add helper method so I agree with @ezr-ondrej. 

Few observations from code changes:
- Both .erb files still having these conditions in place and this method is returning just same class name with taxonomy mapping.
- IIUC, while rendering the form, it gives call to this `tab_classes_for_edit_auth_source_external` method thrice. Every time it is going to execute `show_location_tab?` method. That's why it is important to check what this method does inside?

Either by using any variable at ERB level or if you would like to keep helper then returning the output like
`{ :show_location => show_location_tab?, :show_organization => show_organization_tab?}` would be enough. You can use this value throughout form ERB.

@aaditi-jain, @ezr-ondrej, what do you think?

**Comment by aaditi-jain** [comment]
@kgaikwad :
Thank you for looking into this :)
I am okay to put the entire logic in the view files (like @ezr-ondrej suggested), if putting it into helper seems like a over-kill.

Also about the code changes suggested in the above [comment](https://github.com/theforeman/foreman/pull/7582#pullrequestreview-398492905), here is an explanation: 

From the existing [auth_source_externals/_form.rb](https://github.com/theforeman/foreman/blob/10e7aa5c16283d90dd92d679051ac2853d742069/app/views/auth_source_externals/_form.html.erb#L4) code, it seems that there can be a condition where current_user has show_organization permission and **NOT** show_location permission. In that case, the user should see only an **active** organization tab. This is the condition that is handled in the helper method in this MR. We can also add conditions and tackle it in the view files as @ezr-ondrej suggested.

However, if I updated the helper method to return `{ :show_location => show_location_tab?, :show_organization => show_organization_tab? }` as suggested in the comment, it will require that we add logic in view files also to check the value of :show_location / :show_organization key. That is, the logic will spread both in the helper and in the view files, which kills the purpose of introducing the helper.

Hope that explains the changes. Let me know your thoughts

**Comment by kgaikwad** [comment]
> However, if I updated the helper method to return `{ :show_location => show_location_tab?, :show_organization => show_organization_tab? }` as suggested in the comment, it will require that we add logic in view files also to check the value of :show_location / :show_organization key.

@aaditi-jain,
Yes, it will definitely need changes at view level to check the value by using either variable or helper_method. 
I will prefer a variable to avoid unnecessarily multiple calls to methods `show_location_tab? & show_organization_tab?` and rest of the conditional handling will remain as @ezr-ondrej  suggested.

**Comment by aaditi-jain** [comment]
The more I look at the code, the more I do not see a good reason to move the logic to views from the helper
Here is my reasoning:
1. We are introducing 4 conditions across 2 view files if we move the logic to view. This is reduced to 1 condition statement contained at in a single method by using helper.

1. Considering the "complexity" of the helper method, which was pointed out [earlier](https://github.com/theforeman/foreman/pull/7582#pullrequestreview-398492905) as a reason to shift to views, I found that there are similar helper methods used generously across the code base. Hence, using helpers seems to be consistent with the code pattern followed in the codebase too. Here are examples from existing code base:

- TaxonomyHelper#edit_taxonomy_path
https://github.com/theforeman/foreman/blob/33a5b34b2a984c3689639f3df08f23e83b2104aa/app/helpers/taxonomy_helper.rb#L52
- TaxonomyHelper#mismatches_taxonomies_path
https://github.com/theforeman/foreman/blob/33a5b34b2a984c3689639f3df08f23e83b2104aa/app/helpers/taxonomy_helper.rb#L76
- TemplatesHelper#hide_resource_type_input
https://github.com/theforeman/foreman/blob/33a5b34b2a984c3689639f3df08f23e83b2104aa/app/helpers/templates_helper.rb#L57
- EditorHelper#template_name_attribute
https://github.com/theforeman/foreman/blob/33a5b34b2a984c3689639f3df08f23e83b2104aa/app/helpers/editor_helper.rb#L2

Taking cue from an earlier suggestion, I have introduced a variable in view file to store the output of the existing helper method and updated the code.
@kgaikwad, looking forward to know your thoughts :)

File: app/views/taxonomies/_loc_org_tabs.html.erb
Comment (ezr-ondrej) [inline app/views/taxonomies/_loc_org_tabs.html.erb]: This is equal and safes string interpolation, mainly it's bit more readable IMHO.
```suggestion
  <div class="tab-pane <%= html_options[:tab_classes][:location] %>" id="locations">
```

File: app/views/taxonomies/_loc_org_tabs.html.erb
Comment (ezr-ondrej) [inline app/views/taxonomies/_loc_org_tabs.html.erb]: `^^`

**Review by ezr-ondrej — COMMENTED** [review]
I don't have strong opinion on the helper :)
It works for me eighter way :+1:

File: app/views/auth_source_externals/_form.html.erb
Comment (kgaikwad) [inline app/views/auth_source_externals/_form.html.erb]: External auth source form is having only two tabs (i.e locations & organizations). It seems like locations tab is primary one. 
Can't we just add class `active` directly as to show this tab we already had this condition check?

```suggestion
      <li class="active">
```

**Review by kgaikwad — CHANGES_REQUESTED** [review]
Thank you @aaditi-jain for providing the pointer files and quick updates.
I still vote for changes without a helper method :-) but I am OK with the current changes.

Just added one inline comment.  
To make it consistent with other changes, you can ignore below comment but thought to share it.

File: app/views/auth_source_externals/_form.html.erb
Comment (ezr-ondrej) [reply app/views/auth_source_externals/_form.html.erb]: ~I don't believe this is possible. This template is used in other places and those have more tabs :)~
EDIT: I was too quick I though I'm commenting on different file, this is a good idea :+1:

File: test/helpers/auth_source_external_helper_test.rb
Comment (kgaikwad) [inline test/helpers/auth_source_external_helper_test.rb]:
```suggestion
        assert_empty tab_classes_for_edit_auth_source_external
```

File: test/helpers/auth_source_external_helper_test.rb
Comment (kgaikwad) [inline test/helpers/auth_source_external_helper_test.rb:26]:
```suggestion
        assert_not_empty tab_classes_for_edit_auth_source_external[:organization]
```

File: test/helpers/auth_source_external_helper_test.rb
Comment (kgaikwad) [inline test/helpers/auth_source_external_helper_test.rb:19]:
```suggestion
       assert_not_empty tab_classes_for_edit_auth_source_external[:location]
```

File: test/helpers/auth_source_external_helper_test.rb
Comment (kgaikwad) [inline test/helpers/auth_source_external_helper_test.rb:11]:
```suggestion
         assert_not_empty tab_classes_for_edit_auth_source_external[:location]
```

**Review by kgaikwad — CHANGES_REQUESTED** [review]
Thank you @aaditi-jain!
It would be good to verify using specific key value than comparing the whole object in tests so added small nitpicks. Apart of this everything looks good for first version :+1:

File: test/helpers/auth_source_external_helper_test.rb
Comment (aaditi-jain) [reply test/helpers/auth_source_external_helper_test.rb:19]: @kgaikwad The suggested changes will not cover the method's functionality completely.
We are not only testing here that location tab has 'active' classes, but also that organization tab does not have any 'active' class. 
`assert_not_empty tab_classes_for_edit_auth_source_external[:location]` only checks that :location tab is active. It does not check that the :organization key is not present.

File: test/helpers/auth_source_external_helper_test.rb
Comment (aaditi-jain) [reply test/helpers/auth_source_external_helper_test.rb:11]: Same as https://github.com/theforeman/foreman/pull/7582#discussion_r419672478

File: test/helpers/auth_source_external_helper_test.rb
Comment (aaditi-jain) [reply test/helpers/auth_source_external_helper_test.rb:26]: Similar to as https://github.com/theforeman/foreman/pull/7582#discussion_r419672478

**Comment by aaditi-jain** [comment]
> It would be good to verify using specific key value than comparing the whole object in tests so added small nitpicks. Apart of this everything looks good for first version +1

@kgaikwad I have added my opinion about the suggested changes. Looking forward to know your thoughts on it :). Also not really sure what do we mean by `first version` mentioned above.

File: test/helpers/auth_source_external_helper_test.rb
Comment (kgaikwad) [reply test/helpers/auth_source_external_helper_test.rb:19]: > It does not check that the :organization key is not present.

Suggested this change because for any change in tab_classes, one will need to touch these test cases unnecessarily. What do you think about adding multiple assert statements in one test? Here, for example -

```
assert_equal [:location], tab_classes_for_edit_auth_source_external.keys
assert_not_empty tab_classes_for_edit_auth_source_external[:location]
```

**Review by kgaikwad — APPROVED** [review]
It would be good to avoid hash comparison but it works. 
I don't want to block this PR further due to those changes. 

Thank you @aaditi-jain for quick updates on review comments :ok_hand: :tada:

@rahulbajaj0509, @ezr-ondrej  - any additional comments from your side?

**Review by rabajaj0509 — APPROVED** [review]
I think this one is a good to go! thanks @aaditi-jain

**Review by ezr-ondrej — APPROVED** [review]
I'm ok with this. As this is bugfix, it should get it in before branching.
Right @tbrisker ?

**Review by tbrisker — APPROVED** [review]
Thanks @aaditi-jain and all reviewers!

**Comment by tbrisker** [comment]
2.0 cherry-pick - d98547067

**Comment by aaditi-jain** [comment]
Thank you @kgaikwad, @ezr-ondrej for the reviews :smiley:. Thank you @rahulbajaj0509 too for helping in setting up foreman on my local and in picking up a first issue.

**Comment by alexjfisher** [comment]
This PR seems related to an issue I've discovered after updating to 2.0.1 (from 1.24).  If I'm logged into a specific org/location (such as 'default location'), then I get a 404 when I click on my ldap authentication source.  I've opened a new ticket. 
 See https://projects.theforeman.org/issues/30319

**Pull request merged**
